application/src/main/java/org/togetherjava/tjbot/commands/github/GitHubCommand.java

```diff
@@ -0,0 +1,127 @@
+import org.togetherjava.tjbot.commands.utils.StringDistances;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+
+/**
+ * Slash command (/github) used to search for an issue in one of the repositories listed in the
+ * config. It even comes with auto completion!
+ */
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+public class GitHubCommand extends SlashCommandAdapter {
+    private static final long ONE_MINUTE_IN_MILLIS = 60_000L;
+
+    /**
+     * The command option name of this slash command (this is only there to make the linter happy
+     * lol)
+     */
+    private static final String COMMAND_OPTION_NAME = "title";
+
+    private final GitHubReference reference;
+
+    /**
+     * The last time the cache was updated (millis after Jan 1st 1970)
+     * <p>
+     * The cache updates every minute (see the ONE_MINUTE_IN_MILLIS constant)
+     */
+    private long lastCache;
+    /**
+     * The currently cached auto completion
+     */
+    private List<String> autocompleteCache;
+
+    public GitHubCommand(GitHubReference reference) {
+        super("github", "Search GitHub for an issue", CommandVisibility.GUILD);
+
+        this.reference = reference;
+
+        getData().addOption(OptionType.STRING, COMMAND_OPTION_NAME,
+                "Title of the issue you're looking for", true, true);
+
+        updateCache();
+    }
+
+    @Override
+    public void onSlashCommand(SlashCommandInteractionEvent event) {
+        String title = event.getOption(COMMAND_OPTION_NAME).getAsString();
+        Matcher matcher = GitHubReference.ISSUE_REFERENCE_PATTERN.matcher(title);
+
+        if (!matcher.find()) {
+            event.reply("Could not parse your query").setEphemeral(true).queue();
+
+            return;
+        }
+
+        int id = Integer.parseInt(matcher.group(1));
+        Optional<GHIssue> issue = reference.findIssue(id);
+
+        if (issue.isPresent()) {
+            try {
+                event.replyEmbeds(reference.generateReply(issue.get())).queue();
+            } catch (IOException ex) {
+                throw new UncheckedIOException(ex);
+            }
+        } else {
+            event.reply("Could not find the issue you are looking for").setEphemeral(true).queue();
+        }
+    }
+
+    @Override
+    public void onAutoComplete(CommandAutoCompleteInteractionEvent event) {
+        String title = event.getOption(COMMAND_OPTION_NAME).getAsString();
+
+        if (title.isEmpty()) {
+            event.replyChoiceStrings(autocompleteCache.stream().limit(25).toList()).queue();
+        } else {
+            event
+                .replyChoiceStrings(autocompleteCache.stream()
+                    .sorted(Comparator.comparingInt(s -> StringDistances.editDistance(title,
```

Comment:
sure u want to sort all of them? think about using a `PriorityQueue` with that as `Comparator` and then calling `poll()` `25-times`:
```java
Queue<Foo> foos = new PriorityQueue<>(Comparator...);

List<Foo> foosBla = Stream.generate(foos::poll)
        .limit(25)
        .toList();
```

Reply:
the priority queue would still sort them internally though no?

Reply:
it does k-sort, not full-sort. thats MUCH faster. not that its super crucial here, not like we have thousands of issues.

Reply:
done

Reply:
we have 3 PRs now that essentially duplicated this pattern of "i have a list and want to get the best x matches". we should probably elevate this into a helper somewhere. but alright, lets do that afterwards, once all PRs are merged.